src/main/java/frc/robot/subsystems/Output.java

```diff
@@ -0,0 +1,412 @@
+    switch (offset) {
+      case LEFT:
+        inOffset = -12.94 / 2;
+        break;
+      case RIGHT:
+        inOffset = 12.94 / 2;
+        break;
+      default:
+        break;
+    }
+    Pose2d tagPose = field.getTagPose(TagID).orElse(new Pose3d()).toPose2d();
+    Transform2d poseOffset = new Transform2d(Constants.kRobotWidth.div(2), Inches.of(inOffset),
+        Rotation2d.fromDegrees(270));
+    return tagPose.transformBy(poseOffset);
+  }
+
+  public void LocationService(SwerveDrive Drive) {
+    m_drive = Drive;
+    offsetSub = OffsetTopic.subscribe(0);
+  }
+
+  public enum Offset {
+    LEFT(-1), RIGHT(1), CENTER(0);
+
+    private int val;
+
+    Offset(int i) {
+      val = i;
+    }
+
+    public int getVal() {
+      return val;
+    }
+
+
+
+  }
+
+  private static final AprilTagFieldLayout field =
+      AprilTagFieldLayout.loadField(AprilTagFields.k2025ReefscapeWelded);
+
+  List<AprilTag> tags = field.getTags();
+  private SwerveDrive m_drive;
+  private List<Integer> redSourceTags = List.of(1, 2);
+  private List<Integer> blueSourceTags = List.of(12, 13);
+  // private List<Integer> redReefTags = List.of(6, 7, 8, 9, 10, 11);
+  // private List<Integer> blueReefTags = List.of(17, 18, 19, 20, 21, 22);
+  // private List<Integer> bargeTags = List.of(4, 5, 14, 15);
+
+  NetworkTableInstance inst = NetworkTableInstance.getDefault();
+  NetworkTable table = inst.getTable("SmartDashboard");
+
+  IntegerTopic OffsetTopic = table.getIntegerTopic("Offset");
+  IntegerSubscriber offsetSub;
+
+  // Returns the closest tag ID to the robot
+  public int closestTagId() {
+    Pose2d robot = m_drive.getPose();
+    ListIterator<AprilTag> iter = tags.listIterator();
+    int closestTagId = -1;
+    double closestDistance = Double.MAX_VALUE;
+    while (iter.hasNext()) {
+      AprilTag tag = iter.next();
+      double distance =
+          robot.getTranslation().getDistance(tag.pose.getTranslation().toTranslation2d());
+      if (distance < closestDistance) {
+        closestDistance = distance;
+        closestTagId = tag.ID;
+      }
+    }
+    return closestTagId;
+  }
+
+  public boolean inSourceRegion() {
+    Optional<Alliance> ally = DriverStation.getAlliance();
+    if (ally.isPresent()) {
+      if (ally.get() == Alliance.Red) {
+        return redSourceTags.contains(closestTagId()); // red alliance source april tags
+      }
+      if (ally.get() == Alliance.Blue) {
+
+        return blueSourceTags.contains(closestTagId()); // blue alliance source april tags
+      }
+    }
+    return false;
+  }
+
+  public Pose2d getTagAutoPose2d() {
+    int offsetNum = (int) offsetSub.get();
+    Offset offset = Offset.CENTER;
+    switch (offsetNum) {
+      case -1:
+        offset = Offset.LEFT;
+        break;
+      case 1:
+        offset = Offset.RIGHT;
+        break;
+      default:
+        break;
+    }
+    {
+      return genPoseForSourceFromTag(offsetNum, offset);
+    }
+  }
+
+  public class ReefscapeReefSimulation {
```

Comment:
`org.ironmaple.simulation.seasonspecific.reefscape2025.ReefscapeReefSimulation` The reef simulation class already exists?